nixos/modules/services/networking/pleroma.nix

```diff
@@ -0,0 +1,135 @@
+{ config, options, lib, pkgs, stdenv, ... }:
+let
+  cfg = config.services.pleroma;
+in {
+  options = {
+    services.pleroma = with lib; {
+      enable = mkEnableOption "pleroma";
+
+      package = mkOption {
+        type = types.package;
+        default = pkgs.pleroma-otp;
+        description = "Pleroma package to use.";
+      };
+
+      user = mkOption {
+        type = types.str;
+        default = "pleroma";
+        description = "User account under which pleroma runs.";
+      };
+
+      group = mkOption {
+        type = types.str;
+        default = "pleroma";
+        description = "Group account under which pleroma runs.";
+      };
+
+      configs = mkOption {
+        type = with types; listOf str;
+        description = ''
+          Pleroma public configuration.
+
+          This list gets appended from left to
+          write into /etc/pleroma/config.exs. Elixir evaluates its
+          configuration imperatively, meaning you can override a
+          setting by appending a new str to this NixOS option list.
+
+          <emphasis>DO NOT STORE ANY PLEROMA SECRET
+          HERE</emphasis>, use
+          <link linkend="opt-services.pleroma.secretConfigFile">services.pleroma.secretConfigFile</link>
+          instead.
+
+          This setting is going to be stored in a file part of
+          the Nix store. The Nix store being world-readable, it's not
+          the right place to store any secret
+          '';
+      };
+
+      secretConfigFile = mkOption {
+        type = types.str;
+        default = "/var/lib/pleroma/secrets.exs";
+        description = ''
+          Path to the file containing your secret pleroma configuration.
+
+          <emphasis>DO NOT POINT THIS OPTION TO THE NIX
+          STORE</emphasis>, the store being world-readable, it'll
+          compromise all your secrets.
+        '';
+      };
+    };
+  };
+
+  config = lib.mkIf cfg.enable {
+    users = {
+      users."${cfg.user}" = {
+        description = "Pleroma user";
+        home = "/var/lib/pleroma";
```

Comment:
Please make this configurable.

Reply:
This PR is focussing on systemd, see `StateDirectory` in https://www.freedesktop.org/software/systemd/man/systemd.exec.html for more context.

I think pointing to the service state directory is a good default.

If you're worried about the default home value and really care about it, you can override that in your machine description conf using `mkForce`.

Reply:
I don't care about the home dir, I care about where my data is stored.
The fact that the "StateDirectory" is not meant to be configurable is another terrible idea if systemd fanboys forcing their unjustified opinions upon the world.
`mkForce` does nothing in that regard as it cannot change the root of `StateDirectory`

Reply:
The way I handle those things:

```
fileSystems."/var/lib/service" = {
   fsType = "none";
   options = "bind";
   device = "/somewhere/where/my/data/is/stored";
};
```

While I hate it not being configurable it also ensures that there is a defined set of directories that are being mutated by a daemon (and I do not have to guess or read more docs). If I do not care about the data I do not mount a proper volume there but if I do I usually store it elsewhere.

Reply:
> While I hate it not being configurable it also ensures that there is a defined set of directories that are being mutated by a daemon (and I do not have to guess or read more docs).

Except in practice, it's not like that at all because you basically cannot know where data is stored if you can neither configure it nor want to look at source and docs. So in the end, you'll have to read the source code of at least the module to find out if it actually does what you expect it to do, or roll the dice on whether your backups, storage / disk monitoring is actually doing something or if you're just lucky enough never to notice.

Reply:
No, all services go to `/var/lib/<somewhere>` that is pretty easy to predict. Anything else is just wrong and not predictable. For backups you should be more explicit and there I partially agree with having the path in the module system. You should collect them from the services you care about or just backup all of `/var/lib`. It makes reasoning about a system a lot easier if things are following one rule - be it your or my favorite or not. I am sick of services that want to write to `/opt/somewhere` and others requiring a home dir in `/home` etc..

Reply:
> You should collect them from the services you care about or just backup all of /var/lib

Only backing up all of `/var/lib` is currently an option, which will give you a lot of garbage in addition to the services you actually want, because every service throws their state files in `/var/lib`, even if `/var/cache` or `/var/run` is technically more correct.

Anyways, it's easier to re-implement a service on a flake than to change anything on NixOS and besides, there is no point in having a theoretical discussion about this on a pull request of a specific service.

Reply:
> Only backing up all of /var/lib is currently an option, which will give you a lot of garbage in addition to the services you actually want, because every service throws their state files in /var/lib, even if /var/cache or /var/run is technically more correct.

Let's move back to this PR scope. Only uploads and static files will be stored in `/var/lib/pleroma`.

>  The fact that the "StateDirectory" is not meant to be configurable is another terrible idea if systemd fanboys forcing their unjustified opinions upon the world.

This is actually coming from Linux's FHS spec. You can learn more about it there: https://refspecs.linuxfoundation.org/FHS_3.0/fhs/ch05s08.html.

Feel free to use Andi's trick to store your data elsewhere.

This module won't store any state data in `/var/cache` nor in `/var/run`.

Reply:
I am well aware of the FHS, that is why I know that `static files and uploads` belong to `/srv` and not `/var/lib`, as also stated in the footnote on the page you linked. 
Also quoting the FHS for NixOS is quite ironic in itself I'd argue.

Again, moot point as this will neither convince anyone not should it.

Reply:
I just added a read only option pointing to the statedir, making easier to point to it in your machine description.

```
fileSystems."${config.services.pleroma.stateDir}" = {
   fsType = "none";
   options = "bind";
   device = "/somewhere/where/my/data/is/stored";
};

```